### ISSUE TITLE
cnn 기반 얼굴인식

### ISSUE THREAD

**Issue by xown3197** [xown3197/TIL#1]
• Spatial convolutions that slide a kernel over the input feature maps,
• Linear or fully connected layers that take a weighted sum of all the input units, and
• Pooling that take the max, average, or Euclidean norm over spatial regions
- 활성화 함수 RELU
- 이미지 전처리(normalized, fixed-size) 후 NN에 입력
- NN을 통해 고차원적인 이미지를 저차원적으로 해석
- 해석된 이미지는 Clustering, Classifiers 에 효과적임

### Deepface
- 첫 전처리로 3D 페이스 모델링을 이용하여 다른 각도에서 찍은 얼굴도 정면으로 정상화 시킴
- FC : softmax
- 세 가지 요소로 혁신을 일으킴
```
(a) the 3D alignment
(b) a neural network structure with 120 million parameters, and
(c) training with 4.4 million labeled faces
```
- DeepFace는 여러 사람의 얼굴을 군집화하기 어렵다

### FaceNet
- FaceNet's trplet loss fuction을 통해 직접적으로 군집화가 가능
- unit hypershpere는 원점으로부터 거리 1을 가진 고차원적 원이다. 
- FaceNet의 4가지 혁신적 요소
```
(a) the triplet loss
(b) their triplet selection procedure
(c) training with 100 million to 200 million labeled images
(d) (not discussed here) large-scale experimentation to find an network architecture.
```

### Design
- 신경망 훈련에는 Torch를 사용
- 배열 및 선형대수 연산에는 numpy
- 컴퓨터 비전 기초체에는 OpenCV를 사용
- 분류에는 scikit-learn
- matplotlib으로 플로팅 스크립트 제공
- 신경망 아키텍쳐로는 FaceNet을 사용
- 검출기로는 dlib's pretrained face detection을 사용

### preprocessing : Alignment with an Affine Transformation
- 얼굴 주변을 바운딩 박스를 해서 배열로 변환
- 조명이나 포즈에 대해 정규화를 작업
- OpenFace의 경우 눈과 입의 위치를 기준으로 정면을 바라보도록함
- dlib의 얼굴 검출기를 이용하여 2D Affine하여 정면을 바라보도록함
- resize 96 x 96 이미지로 변환

### Training the Face Representation Neural Network
- FaceNet = 100M~200M 이미지로 미리 학습
- DeepFace = 4.4M
- OpenFace = 500k img (combined CASIA-WebFace, FaceScrub)
- OpenFace NN = FaceNet's nn4 = based on GoogLeNet
- Using function
```
a. FaceNet's triplet loss <- unit hypersphere
b. Euclidean distances는 유사성에 사용
```
- 15명의 사람당 20장의 이미지를 사용함
